services/containerregistry/containerregistry.go

```diff
@@ -0,0 +1,48 @@
+package containerregistry
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/manicminer/hamilton/auth"
+)
+
+// ContainerRegistryClient handles communication with Azure Container Registry
+type ContainerRegistryClient struct {
+	authorizer auth.Authorizer
+	httpClient *http.Client
+	serverURL  string
+	tenantID   string
+}
+
+// NewContainerRegistryClient returns a ContainerRegistryClient
+func NewContainerRegistryClient(authorizer auth.Authorizer, serverURL string, tenantID string) *ContainerRegistryClient {
+	httpClient := http.DefaultClient
+	return &ContainerRegistryClient{
+		authorizer,
+		httpClient,
+		serverURL,
+		tenantID,
+	}
+}
+
+// WithHttpClient replaces what http client is used for communication to Azure Container Registry
+func (c *ContainerRegistryClient) WithHttpClient(httpClient *http.Client) {
```

Comment:
I think it's fine to support user-supplied clients, I'm just curious if you have you found a good use for this other than mocking or debugging?

Reply:
Ah, good! 👍 I'm using it here for mocking. Two use cases are tracing / logging and additional authentication (like client certificate). Don't know how useful those are in this case, but seems like something one should do to be a bit more pluggable for custom use cases.